Source/JavaScriptCore/runtime/SyntheticModuleRecord.cpp

```diff
@@ -0,0 +1,133 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "SyntheticModuleRecord.h"
+
+#include "BuiltinNames.h"
+#include "JSCInlines.h"
+#include "JSInternalPromise.h"
+#include "JSModuleEnvironment.h"
+#include "JSModuleNamespaceObject.h"
+#include "JSONObject.h"
+
+namespace JSC {
+
+const ClassInfo SyntheticModuleRecord::s_info = { "ModuleRecord"_s, &Base::s_info, nullptr, nullptr, CREATE_METHOD_TABLE(SyntheticModuleRecord) };
+
+
+Structure* SyntheticModuleRecord::createStructure(VM& vm, JSGlobalObject* globalObject, JSValue prototype)
+{
+    return Structure::create(vm, globalObject, prototype, TypeInfo(ObjectType, StructureFlags), info());
+}
+
+SyntheticModuleRecord* SyntheticModuleRecord::create(JSGlobalObject* globalObject, VM& vm, Structure* structure, const Identifier& moduleKey)
+{
+    SyntheticModuleRecord* instance = new (NotNull, allocateCell<SyntheticModuleRecord>(vm)) SyntheticModuleRecord(vm, structure, moduleKey);
+    instance->finishCreation(globalObject, vm);
+    return instance;
+}
+
+SyntheticModuleRecord::SyntheticModuleRecord(VM& vm, Structure* structure, const Identifier& moduleKey)
+    : Base(vm, structure, moduleKey)
+{
+}
+
+void SyntheticModuleRecord::destroy(JSCell* cell)
+{
+    SyntheticModuleRecord* thisObject = static_cast<SyntheticModuleRecord*>(cell);
+    thisObject->SyntheticModuleRecord::~SyntheticModuleRecord();
+}
+
+void SyntheticModuleRecord::finishCreation(JSGlobalObject* globalObject, VM& vm)
+{
+    Base::finishCreation(globalObject, vm);
+    ASSERT(inherits(info()));
+}
+
+template<typename Visitor>
+void SyntheticModuleRecord::visitChildrenImpl(JSCell* cell, Visitor& visitor)
+{
+    SyntheticModuleRecord* thisObject = jsCast<SyntheticModuleRecord*>(cell);
+    ASSERT_GC_OBJECT_INHERITS(thisObject, info());
+    Base::visitChildren(thisObject, visitor);
+}
+
+DEFINE_VISIT_CHILDREN(SyntheticModuleRecord);
+
+Synchronousness SyntheticModuleRecord::link(JSGlobalObject*, JSValue)
+{
+    return Synchronousness::Sync;
+}
+
+JSValue SyntheticModuleRecord::evaluate(JSGlobalObject*)
+{
+    return jsUndefined();
+}
+
```

Comment:
To add support for arbitrary export names and values, what about something like this? I haven't tried compiling/running this code
```suggestion
SyntheticModuleRecord* SyntheticModuleRecord::tryCreateWithExportNamesAndValues(JSGlobalObject* globalObject, const Identifier& moduleKey, Vector<Identifier>& exportNames, Vector<JSValue>& exportValues)
{
    VM& vm = globalObject->vm();
    auto scope = DECLARE_THROW_SCOPE(vm);

    auto* moduleRecord = create(globalObject, vm, globalObject->syntheticModuleRecordStructure(), moduleKey);

    SymbolTable* exportSymbolTable = SymbolTable::create(vm);
     {
        auto offset = exportSymbolTable->takeNextScopeOffset(NoLockingNecessary);
        exportSymbolTable->set(NoLockingNecessary, vm.propertyNames->starNamespacePrivateName.impl(), SymbolTableEntry(VarOffset(offset)));
    }

    ASSERT(exportNames.size() == exportValues.size());

    for (size_t i = 0; i < exportNames.size(); ++i) {
        auto exportName = exportNames[i];
        auto exportValue = exportValues[i];
       
        moduleRecord->addExportEntry(ExportEntry::createLocal(exportName, exportName));
        auto offset = exportSymbolTable->takeNextScopeOffset(NoLockingNecessary);
        exportSymbolTable->set(NoLockingNecessary, exportName, SymbolTableEntry(VarOffset(offset)));
    }

    JSModuleEnvironment* moduleEnvironment = JSModuleEnvironment::create(vm, globalObject, nullptr, exportSymbolTable, jsTDZValue(), moduleRecord);
    moduleRecord->setModuleEnvironment(globalObject, moduleEnvironment);
    RETURN_IF_EXCEPTION(scope, { });

    bool shouldThrowReadOnlyError = false;
    bool ignoreReadOnlyErrors = true;
    bool putResult = false;

    for (size_t i = 0; i < exportNames.size(); ++i) {
        JSC::PropertyName exportName(exportNames[i]);
        auto exportValue = exportValues[i];       
        putResult = symbolTablePutTouchWatchpointSet(moduleEnvironment, globalObject, exportName, exportValue, shouldThrowReadOnlyError, ignoreReadOnlyErrors, putResult);
        RETURN_IF_EXCEPTION(scope, { });
        RELEASE_ASSERT(putResult);
    }
    RETURN_IF_EXCEPTION(scope, { });

    return moduleRecord;
}

SyntheticModuleRecord* SyntheticModuleRecord::createWithExportNamesAndValues(JSGlobalObject* globalObject, const Identifier& moduleKey, Vector<Identifier>& exportNames, Vector<JSValue>& exportValues)
{
    VM& vm = globalObject->vm();
    auto scope = DECLARE_THROW_SCOPE(vm);
    RELEASE_AND_RETURN(scope, SyntheticModuleRecord::tryCreateDefaultExportSyntheticModule(globalObject, moduleKey, result));    
}
```

Reply:
That sounds good.